### PR TITLE
Proper themes (Fixes #56)

### LINKED ISSUE #56: Implement dynamic color themes
#### General information
- **Device:** doesn't matter
- **Android Version:** doesn't matter
- **Pod:** doesn't matter
- **Diaspora pod version:** doesn't matter
- **App source:** doesn't matter
- **App version:** HEAD

It would be nice to have an option where the user can dynamically chose the colors for the theme. [LeafPic](https://github.com/HoraApps/LeafPic) is a very good example for that:
## 

Todo: 
- [x] Create color picker in settingsActivity
- [x] Store color values in AppSettings
- [x] Apply colors to MainActivity
- [x] Apply colors to AboutActivity
- [x] Apply colors to DiasporaStreamFragment
- [x] Fix: StatusBar background color is still blue when clicking on a PreferenceScreen in the Settings
- ~~Create/apply light/dark theme (fonts, icons, background of drawer/Activities)~~
- ~~Apply accent colors to checkboxes etc. in the SettingsActivity~~

### PR DESCRIPTION
**Pull request by vanitasvitae**

### CODE REVIEW

**Review by vanitasvitae — COMMENTED** [review]
There are some bugs on older versions of Android (4.4). I'll have to investigate that further.

File: app/src/main/java/com/github/dfa/diaspora_android/util/ThemeHelper.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        }
+    }
+
+    public static int getPrimaryDarkColor() {
+        return ColorPalette.getObscuredColor(getPrimaryColor());
+    }
+
+    public static void updateActionBarColor(ActionBar actionBar) {
+        if(actionBar != null) {
+            actionBar.setBackgroundDrawable(new ColorDrawable(getInstance().appSettings.getPrimaryColor()));
```
Comment (gsantner) [inline app/src/main/java/com/github/dfa/diaspora_android/util/ThemeHelper.java]: Do we need ActionBar? We only use ToolBar everywhere :)

File: app/src/main/res/layout/color_picker__dialog.xml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:orientation="vertical"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+    <FrameLayout
+        android:id="@+id/color_picker_title_background"
```
Comment (gsantner) [inline app/src/main/res/layout/color_picker__dialog.xml]: `color_picker__dialog__title_background`
`color_picker__dialog__` as prefix

File: app/src/main/res/layout/color_picker__dialog_title.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/color_picker_title_background"
+    android:orientation="vertical" android:layout_width="match_parent"
+    android:layout_height="match_parent">
+    <TextView
+        android:id="@+id/color_picker_title"
```
Comment (gsantner) [inline app/src/main/res/layout/color_picker__dialog_title.xml]: `color_picker__dialog_title__color_picker_title`
`color_picker__dialog_title__` as prefix

File: app/src/main/res/values-de/strings-preferences.xml [hunk 4]
```diff
@@ -14,6 +14,13 @@
   <string name="pref_title__sub_nav_slider">Navigations Slider</string>
   <string name="pref_desc__sub_nav_slider">Konfiguration der Sichtbarkeit von Einträgen im Navigation-Slider</string>
   <string name="pref_cat__visibility_nav_items">Sichtbarkeit der Einträge</string>
+  <!-- Themes -->
```
Comment (gsantner) [inline app/src/main/res/values-de/strings-preferences.xml:17]: Do not check in translations (other than `values`). Will get overwritten by Crowdin :)

File: app/src/main/res/values/color.xml [hunk 5]
```diff
@@ -5,10 +5,7 @@
     <color name="colorAccent">@color/accent</color>
 
     <!-- Colors from Palette -->
-    <color name="primary">#207be6</color>
-    <color name="primary_dark">#195ed2</color>
     <color name="primary_light">#BBDEFB</color>
-    <color name="accent">#FF5300</color>
```
Comment (gsantner) [inline app/src/main/res/values/color.xml:11]: Maybe keep color `accent,primary,primary_dark` (for the default values) and reference the color behind it
`<color name="accent">@color/md_deep_orange_500</color>`

File: app/src/main/res/values/strings-preferences.xml [hunk 6]
```diff
@@ -16,6 +16,15 @@
     <string name="pref_key__clear_cache" translatable="false">pref_key_clear_cache</string>
     <string name="pref_key__chrome_custom_tabs_enabled" translatable="false">pref_key__chrome_custom_tabs_enabled</string>
 
+    <!-- Themes -->
+    <string name="pref_key__primary_color" translatable="false">pref_key_primary_color</string> <!--onPreferenceClick-->
```
Comment (gsantner) [inline app/src/main/res/values/strings-preferences.xml]: Remove comment, will go into a **seperate** line in every language when updating from crowdin:

```
<!--onPreferenceClick-->
<string name="pref_key__primary_color_shade" translatable="false">pref_key_primary_color_shade</string>
<string name="pref_key__accent_color" translatable="false">pref_key_accent_color</string>
<!--onPreferenceClick-->
<string name="pref_key__accent_color_base" translatable="false">pref_key_accent_color_base</string>
```

Maybe adjust the key name instead

**Review by gsantner — COMMENTED** [review]
Good job, added comments

File: app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java [hunk 7]
```diff
@@ -100,10 +108,26 @@ public void onClick(View v) {
         }
     }
 
+    @Override
+    protected AppSettings getAppSettings() {
+        return ((App) getApplication()).getSettings();
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java]: Done

File: app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java [hunk 8]
```diff
@@ -100,10 +108,26 @@ public void onClick(View v) {
...
+        ThemeHelper.setPrimaryColorAsBackground(linearLayout);
+    }
+
     /**
      * Fragment that shows general information about the app
      */
-    public static class AboutFragment extends Fragment {
+    public static class AboutFragment extends ThemedFragment {
+
+        TextView aboutText;
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java:150]: Done

File: app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java [hunk 9]
```diff
@@ -112,8 +136,8 @@ public AboutFragment() {
         public View onCreateView(LayoutInflater inflater, ViewGroup container,
                                  Bundle savedInstanceState) {
             View rootView = inflater.inflate(R.layout.about__fragment_about, container, false);
-            TextView appVersion = (TextView) rootView.findViewById(R.id.fragment_about__app_version);
-
+            appVersion = (TextView) rootView.findViewById(R.id.fragment_about__app_version);
+            aboutText = (TextView) rootView.findViewById(R.id.fragment_about__about_text);
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java]: Done

File: app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java [hunk 10]
```diff
@@ -125,18 +149,40 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
...
         }
+
+        @Override
+        protected void applyColorToViews() {
+            ThemeHelper.updateTextViewColor(aboutText);
+        }
+
+        @Override
+        public String getFragmentTag() {
+            return null;
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java]: Done

File: app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java [hunk 11]
```diff
@@ -187,6 +231,28 @@ public String getLicense3dPartyHtml(Context context) {
...
+        @Override
+        protected void applyColorToViews() {
+            ThemeHelper.getInstance(getAppSettings());
+            ThemeHelper.updateTextViewColor(textLicense3partyBox);
+            ThemeHelper.updateTextViewColor(textLicenseBox);
+        }
+
+        @Override
+        public String getFragmentTag() {
+            return null;
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/AboutActivity.java]: Done, but unnecessary (We don't handle these Fragments in the FragmentManager anyways).

File: app/src/main/java/com/github/dfa/diaspora_android/activity/MainActivity.java [hunk 12]
```diff
@@ -967,9 +992,23 @@ public String getTextToBeShared() {
...
+     *
      * @param textToBeShared
      */
     public void setTextToBeShared(String textToBeShared) {
         this.textToBeShared = textToBeShared;
     }
+
+    @Override
+    protected AppSettings getAppSettings() {
+        return (appSettings != null ? appSettings : ((App)getApplication()).getSettings());
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/MainActivity.java]: Done :)

File: app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java [hunk 13]
```diff
@@ -67,6 +81,17 @@ private void setActivityRestartRequired() {
         this.activityRestartRequired = true;
     }
 
+    @Override
+    protected AppSettings getAppSettings() {
+        return ((App)getApplication()).getSettings();
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java]: Done

File: app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java [hunk 14]
```diff
@@ -67,6 +81,17 @@ private void setActivityRestartRequired() {
...
+    }
+
+    @Override
+    protected void applyColorToViews() {
+        ThemeHelper.getInstance(getAppSettings());
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java]: Done in ThemedActivity

File: app/src/main/res/layout/color_picker__dialog.xml (hunk 2 above)
Comment (vanitasvitae) [reply app/src/main/res/layout/color_picker__dialog.xml]: Done

File: app/src/main/res/layout/color_picker__dialog_title.xml (hunk 3 above)
Comment (vanitasvitae) [reply app/src/main/res/layout/color_picker__dialog_title.xml]: Done

File: app/src/main/res/values-de/strings-preferences.xml (hunk 4 above)
Comment (vanitasvitae) [reply app/src/main/res/values-de/strings-preferences.xml:17]: Okay, I'll still leave this in, alright?

File: app/src/main/res/values/strings-preferences.xml (hunk 6 above)
Comment (vanitasvitae) [reply app/src/main/res/values/strings-preferences.xml]: Done

File: app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java [hunk 15]
```diff
@@ -30,23 +32,35 @@
...
+import uz.shift.colorpicker.LineColorPicker;
+import uz.shift.colorpicker.OnColorChangedListener;
 
 /**
  * @author vanitas
  */
-public class SettingsActivity extends AppCompatActivity {
+public class SettingsActivity extends ThemedActivity {
     private boolean activityRestartRequired;
+    private ActionBar toolbar;
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java]: The problem is, that we have no layout defined for the SettingsActivity. If I do, there are two bars (Toolbar from layout resource and Actionbar from PreferenceFragment).

File: app/src/main/java/com/github/dfa/diaspora_android/util/ThemeHelper.java (hunk 1 above)
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/util/ThemeHelper.java]: Used in SettingsActivity

File: app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (gsantner) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java]: How does SettingsActivity look on older devices?

File: app/src/main/res/values-de/strings-preferences.xml (hunk 4 above)
Comment (gsantner) [reply app/src/main/res/values-de/strings-preferences.xml:17]: Yeah, but will get stripped away after first update if not published to crowdin :D

File: app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/activity/SettingsActivity.java]: On my tablet (4.4), it looks just like on any other device. The only slight problem is, that the actionbar does not get recolored there immediately.

File: app/src/main/res/values-de/strings-preferences.xml (hunk 4 above)
Comment (vanitasvitae) [reply app/src/main/res/values-de/strings-preferences.xml:17]: Maybe some translator finds better translations :)

File: app/src/main/java/com/github/dfa/diaspora_android/ui/theming/ColorPalette.java [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+import android.graphics.Color;
+import android.support.v4.content.ContextCompat;
+import android.support.v4.graphics.ColorUtils;
+
+import com.github.dfa.diaspora_android.R;
+
+/**
+ * Created by dnld on 24/02/16.
+ */
+public class ColorPalette {
```
[the remark this answers is not in the document]
Comment (vanitasvitae) [reply app/src/main/java/com/github/dfa/diaspora_android/ui/theming/ColorPalette.java]: I looked around for some time, but I couldn't come up with something that makes sense. What exactly do you have in mind?

**Comment by SansPseudoFix** [comment]
Is it possible to have some screenshots of this? As a translator, it's kind of tricky to guess the context (even with crowdin's help) :)

**Comment by vanitasvitae** [comment]
I can see, if I can upload some screenshots later. For the context: The primary color is the color of the toolbars etc (blue in the last released version). The accemt color is the color of the progressbar and clickable links in the about section of the app. Both colors are now customizable via color sliders. The slider looks like this one from LeafPic: https://github.com/HoraApps/LeafPic/blob/master/screenshots/2.png?raw=true

I hope that made everything a little more clear. Let me know if you are interested in a developer preview build of the app, to see exactly, what changes were made :)

**Comment by SansPseudoFix** [comment]
Actually, I figured out for this one :P 
It's much about notifications dropdown and stuff, here (I thought this PR was about it too). Depend of the context, I have to use the correct formulation, gender...

> Let me know if you are interested in a developer preview build of the app, to see exactly, what changes were made :)

If it's just an `.apk`, yeah, sure!

**Comment by vanitasvitae** [comment]
@SansPseudoFix There you go! https://vanitasvitae.no-ip.biz/nextcloud/index.php/s/2iLufCer8fdEp0j
Happy translating :)

**Pull request merged**